app/scripts/controllers/user-storage/encryption/cache.ts

```diff
@@ -0,0 +1,76 @@
+import { base64ToByteArray, byteArrayToBase64 } from './utils';
+
+type CachedEntry = {
+  salt: Uint8Array;
+  base64Salt: string;
+  key: Uint8Array;
+};
+
+/**
+ * In Memory Cache of the derived key from a given salt.
+ * We can tidy this to make it reusable with future versions and KDF functions
+ */
+const inMemCachedKDF: Map<string, Uint8Array> = new Map();
+
+/**
+ *
+ * @param salt provide salt to receive cached key
+ * @returns cached key
+ */
+export function getCachedKeyBySalt(salt: Uint8Array): CachedEntry | undefined {
+  const base64Salt = byteArrayToBase64(salt);
+  const cachedKey = inMemCachedKDF.get(base64Salt);
+  if (!cachedKey) {
+    return undefined;
+  }
+
+  return {
+    salt,
+    base64Salt,
+    key: cachedKey,
+  };
+}
+
+/**
+ *
+ * @returns any (the first) cached key
+ */
+export function getAnyCachedKey(): CachedEntry | undefined {
+  const cachedEntry: [string, Uint8Array] | undefined = inMemCachedKDF
+    .entries()
+    .next().value;
```

Comment:
(could) explain this construct in a comment. Not everyone remembers what `Map.prototype.entries` returns.

Reply:
Yep this is sweet! Nice use of iterators!

Future - once full browser support, this could be replaced with [Iterator.prototype.take()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/take) - E.g. `inMemCachedKDF.entries().take(1)`